ironfish/src/indexer/minedBlocksIndexer.ts

```diff
@@ -0,0 +1,268 @@
+import { Accounts } from '../account'
+import { Assert } from '../assert'
+import { Blockchain } from '../blockchain'
+import { ChainProcessor } from '../chainProcessor'
+import { FileSystem } from '../fileSystems'
+import { createRootLogger, Logger } from '../logger'
+import { IDatabase, IDatabaseStore, JsonEncoding, StringEncoding } from '../storage'
+import { createDB } from '../storage/utils'
+import { SetTimeoutToken } from '../utils'
+import { isBlockMine } from '../utils/blockchain'
+
+const DATABASE_VERSION = 1
+
+type MinedBlocksDBMeta = {
+  headHash: string | null
+}
+
+type MinedBlock = {
+  main: boolean
+  hash: string
+  sequence: number
+  account: string
+  minersFee: number
+}
+
+export class MinedBlocksIndexer {
+  protected meta: IDatabaseStore<{
+    key: keyof MinedBlocksDBMeta
+    value: MinedBlocksDBMeta[keyof MinedBlocksDBMeta]
+  }>
+  protected minedBlocks: IDatabaseStore<{ key: string; value: MinedBlock }>
+  protected accountToBlockHashes: IDatabaseStore<{ key: string; value: string[] }>
+
+  protected files: FileSystem
+  protected database: IDatabase
+  protected location: string
+  protected readonly accounts: Accounts
+  protected readonly logger: Logger
+  protected isOpen: boolean
+  protected isStarted: boolean
+  protected eventLoopTimeout: SetTimeoutToken | null = null
+  protected chain: Blockchain
+  protected chainProcessor: ChainProcessor
+
+  constructor({
+    files,
+    location,
+    accounts,
+    chain,
+    logger = createRootLogger(),
+  }: {
+    files: FileSystem
+    location: string
+    accounts: Accounts
+    chain: Blockchain
+    logger?: Logger
+  }) {
+    this.files = files
+    this.location = location
+    this.database = createDB({ location })
+    this.accounts = accounts
+    this.logger = logger
+    this.chain = chain
+    this.isOpen = false
+    this.isStarted = false
+
+    this.meta = this.database.addStore<{
+      key: keyof MinedBlocksDBMeta
+      value: MinedBlocksDBMeta[keyof MinedBlocksDBMeta]
+    }>({
+      name: 'minedBlocksDBMeta',
+      keyEncoding: new StringEncoding<keyof MinedBlocksDBMeta>(),
+      valueEncoding: new JsonEncoding(),
+    })
+
+    this.minedBlocks = this.database.addStore<{ key: string; value: MinedBlock }>({
+      name: 'minedBlocks',
+      keyEncoding: new StringEncoding(),
+      valueEncoding: new JsonEncoding(),
+    })
+
+    this.accountToBlockHashes = this.database.addStore<{ key: string; value: string[] }>({
+      name: 'accountToBlockHashes',
+      keyEncoding: new StringEncoding(),
+      valueEncoding: new JsonEncoding(),
+    })
+
+    this.chainProcessor = new ChainProcessor({ logger, chain, head: null })
+
+    this.chainProcessor.onAdd.on(async (header) => {
+      const block = await this.chain.getBlock(header)
+      Assert.isNotNull(block)
+
+      const account = this.accounts.listAccounts().find((a) => isBlockMine(block, a))
+      await this.database.transaction(async (tx) => {
```

Comment:
I think there's a few things here worth thinking through

1. It's possible to start a transaction here that does nothing in the current code
2. There's a data consistency issue with this transaction.